SecurityPkg/Library/AuthVariableLib/AuthService.c

```diff
@@ -1818,6 +1946,37 @@ CleanCertsFromDb (
   return Status;
 }
 
+/**
+  Find hash algorithm index.
+
+  @param[in]  SigData      Pointer to the PKCS#7 message.
+  @param[in]  SigDataSize  Length of the PKCS#7 message.
+
+  @retval UINT8        Hash Algorithm Index.
+**/
+UINT8
+FindHashAlgorithmIndex (
+  IN     UINT8   *SigData,
+  IN     UINT32  SigDataSize
+  )
+{
+  UINT8  i;
+
+  for (i = 0; i < (sizeof (mHashInfo) / sizeof (EFI_HASH_INFO)); i++) {

```

Comment:
Think it's worth cleaning up this function in my patch? The downside is EDK2 copy and pasted this function to two different files in the original commit. 


```c
//
// OID will be at offset 13 if there is no ContentInfo, and at offset 32 if there is ContentInfo.
//
#define ASN1_OID_OFFSET(a, INCLUDES_CONTENT_INFO)  (13 + (INCLUDES_CONTENT_INFO ? 19 : 0) + a)
STATIC_ASSERT (ASN1_OID_OFFSET (0, FALSE) == 13, "ASN1_OID_OFFSET calculation error");
STATIC_ASSERT (ASN1_OID_OFFSET (0, TRUE) == 32, "ASN1_OID_OFFSET calculation error");

//
// Two byte encode will be at offset 1 if there is no ContentInfo, and at offset 20 if there is ContentInfo.
//
#define ASN1_TWO_BYTE_ENCODE_OFFSET(a, INCLUDES_CONTENT_INFO)  (1 + (INCLUDES_CONTENT_INFO ? 19 : 0) + a)
STATIC_ASSERT (ASN1_TWO_BYTE_ENCODE_OFFSET (0, FALSE) == 1, "ASN1_TWO_BYTE_ENCODE_OFFSET calculation error");
STATIC_ASSERT (ASN1_TWO_BYTE_ENCODE_OFFSET (0, TRUE) == 20, "ASN1_TWO_BYTE_ENCODE_OFFSET calculation error");

//
// This is the mask to check if the OID is two byte encoded.
//
#define ASN1_VERIFY_TWO_BYTE_ENCODE(a, INCLUDES_CONTENT_INFO)  ((*(ASN1_TWO_BYTE_ENCODE_OFFSET(a, INCLUDES_CONTENT_INFO)) & TWO_BYTE_ENCODE) == TWO_BYTE_ENCODE)


/**
  Find hash algorithm index.

  @param[in]  SigData      Pointer to the PKCS#7 message.
  @param[in]  SigDataSize  Length of the PKCS#7 message.
  @param[out] Index        Pointer to the hash algorithm index.

  @retval EFI_INVALID_PARAMETER the input parameter is invalid.
**/
EFI_STATUS
FindHashAlgorithmIndex (
  IN     UINT8   *SigData,
  IN     UINT32  SigDataSize,
  OUT    UINT8   *Index
  )
{
  UINT8  i;

  if ((SigData == NULL) || (SigDataSize == 0)) {
    //
    // This function expects valid input.
    // This means that the caller must provide at least enough data to determine the hash algorithm.
    //
    return EFI_INVALID_PARAMETER;
  }

  for (i = 0; i < (sizeof (mHashInfo) / sizeof (EFI_HASH_INFO)); i++) {
    //
    // Verify the length of the PKCS#7 message is at least enough to determine the hash algorithm.
    //
    if (SigDataSize >= (ASN1_OID_OFFSET (mHashInfo[i].OidLength, FALSE))) {
      //
      // The DER-encoded ContentInfo structure is not present.
      //
      if (ASN1_VERIFY_TWO_BYTE_ENCODE (SigData, FALSE) &&
          (CompareMem (ASN1_OID_OFFSET (SigData, FALSE), mHashInfo[i].OidValue, mHashInfo[i].OidLength) == 0))
      {
        *Index = i;
        return EFI_SUCCESS;
      }
    } else if (SigDataSize >= (ASN1_OID_OFFSET (mHashInfo[i].OidLength, TRUE))) {
      //
      // The DER-encoded ContentInfo structure is present.
      //
      if (ASN1_VERIFY_TWO_BYTE_ENCODE (SigData, TRUE) &&
          (CompareMem (ASN1_OID_OFFSET (SigData, TRUE), mHashInfo[i].OidValue, mHashInfo[i].OidLength) == 0))
      {
        *Index = i;
        return EFI_SUCCESS;
      }
    }
  }

  return EFI_UNSUPPORTED;
}
```